### ISSUE TITLE
Think about long term archiving of basic job history

### ISSUE THREAD

**Issue by evansd** [opensafely/job-server#126]
The history of which jobs were run when, by whom, and against what commit is going to become an important thing to preserve. I'm wondering it, as well as storing it in its local database, the job-server should log the key information in some plain text format which can get committed to git and pushed to github. This also means that the entire history gets cryptographically verified for us, which is nice.

**Comment by evansd** [comment]
A CSV file per week would probably do the trick.

**Comment by ghickman** [comment]
This could even be a scheduled Action on the destination repo which hits an endpoint on job-server?  That way we leave the scheduling to GitHub and get logging for free.

**Comment by sebbacon** [comment]
I like this idea as a public backup. It's worth being clear this isn't a substitute for getting our own backups in order, which we should do at some point. Getting very close to needing that roadmap / release schedule I think!

CSVs feels maybe too arbitrary in terms of storage granularity - how about one file per Job + JobRequest?

**Comment by evansd** [comment]
Yes, I agree: it's a different thing from db backups, but also important I think.

> how about one file per Job + JobRequest?
I think this will result in too many small files, which comes with it's own set of problems. Particularly in git.

I like the idea that CSVs would be readable directly on github, but if we're not bothered about that then I think my favourite storage format is gzipped, newline-separated JSON. Flexible, compact (after gzip), but easy to read in any language, and easy to work with in the shell.

**Comment by sebbacon** [comment]
Github pros:
* Covenient location
* CSV rendering
* Search
* Free hosting

Github cons:
* 10Gb limit
* Slight abuse of system as we won't be using any git features
* Requires us to divide logs into daily chunks which is a bit arbitrary

The cons are pretty minor really. Back of an envelope calculation says we're unlikely to hit more than 100Mb per year of logs. So it'll do as an archive for the foreseeable future.

I'm in favour of a django command that dumps the smallest useful subset of job log data to CSV and commits to a repo (perhaps over github api?)

**Comment by evansd** [comment]
For me the other pro of GitHub is that it makes it trivially easy for people to grab a copy of the history and confirm that we don't alter it.

As for getting it into github, I wonder if we could expose the logs over HTTP and then have a GitHub action which pulls them in and commits any changes. That way it becomes a standalone thing that the job-server doesn't even need to know about.

**Comment by ghickman** [comment]
As an initial archive I think we should drop the 114 Jobs which have no JobRequest, the latest of which was created on 16th October.  That would let me tidy up some of the code paths, specifically making Job.job_request a non nullable field.

SQLite can export a query to CSV which seems like a reasonable way to start this off?  I can manually make the CSV, create the archive repo, commit the CSV there, remove the 114 Jobs from the database.

How does that sound as a first step?

**Comment by evansd** [comment]
Sounds good to me!

**Comment by ghickman** [comment]
First export done with: ```
sqlite> .headers on
sqlite> .mode csv
sqlite> .output export.csv
sqlite> select * from jobserver_job where job_request_id is null;
```

Output copied into the [jobs-archive](https://github.com/opensafely/jobs-archive) repo.

**Comment by evansd** [comment]
Fine to close this if there's no intention of ever doing it, but just to note that the `event-log` page is not an archive in the sense originally meant: it's mutable and relies on the code and database to regenerate it. The idea of an archive was to be something static that would be permanently preserved without us needing to do anything special.

**Comment by mikerkelly** [comment]
@evansd thanks.

No opinion on whether we will ever do this so reopening.

**Comment by mikerkelly** [comment]
Relevant Slack threads [here](https://bennettoxford.slack.com/archives/C069YDR4NCA/p1759855820855249) and [here](https://bennettoxford.slack.com/archives/C093H4M6TE2/p1760028696309899).
